core/shared/src/main/scala/scalaz/zio/ZManaged.scala

```diff
@@ -182,6 +182,17 @@ final case class ZManaged[-R, +E, +A](reserve: ZIO[R, E, Reservation[R, E, A]])
   final def either: ZManaged[R, Nothing, Either[E, A]] =
     fold(Left[E, A], Right[E, A])
 
+  /**
+   * Ensures that `f` is executed when this ZManaged is finalized, after
+   * the existing finalizer.
+   */
+  final def ensuring[R1 <: R](f: ZIO[R1, Nothing, _]): ZManaged[R1, E, A] =
+    ZManaged {
+      reserve.map { r =>
+        r.copy(release = r.release *> f)
```

Comment:
Should it be `r.release.ensuring(f)` in case there's a defect in `release`?

Reply:
good catch!! there's a bug in `flatMap` which causes finalizers with defects to prevent earlier finalizers from running.